contracts/contracts/strategies/BaseCurveStrategy.sol

```diff
@@ -2,30 +2,89 @@
 import { StableMath } from "../utils/StableMath.sol";
 import { Helpers } from "../utils/Helpers.sol";
+import { CurveThreeCoinLib } from "./curve/CurveThreeCoinLib.sol";
 
 abstract contract BaseCurveStrategy is InitializableAbstractStrategy {
     using StableMath for uint256;
     using SafeERC20 for IERC20;
 
-    uint256 internal constant MAX_SLIPPAGE = 1e16; // 1%, same as the Curve UI
-    // number of assets in Curve 3Pool (USDC, DAI, USDT)
-    uint256 internal constant THREEPOOL_ASSET_COUNT = 3;
-    address internal pTokenAddress;
+    uint256 public constant MAX_SLIPPAGE = 1e16; // 1%, same as the Curve UI
+    /// @notice number of assets in base Curve pool. eg 3 for the 3Pool
+    uint256 public immutable CURVE_BASE_ASSETS;
+    address public immutable CURVE_POOL;
+    address public immutable CURVE_LP_TOKEN;
+
+    // Only supporting up to 3 coins for now.
+    // The new Stableswap pools support up to 8 coins, but
+    // we'll add support for more than 3 pools later if needed.
+    address public immutable coin0;
+    address public immutable coin1;
+    address public immutable coin2;
+    uint256 public immutable decimals0;
+    uint256 public immutable decimals1;
+    uint256 public immutable decimals2;
+
+    // slither-disable-next-line constable-states
+    address private deprecated_pTokenAddress;
 
     int256[49] private __reserved;
 
+    struct CurveConfig {
+        uint256 curveBaseAssets;
+        address curvePool;
+        address curveLpToken;
+    }
+
+    constructor(CurveConfig memory _curveConfig) {
+        // Only support Curve pools with 2 or 3 coins for now.
+        require(
+            _curveConfig.curveBaseAssets == 2 ||
+                _curveConfig.curveBaseAssets == 3,
+            "Invalid Curve base assets"
+        );
+        require(_curveConfig.curvePool != address(0), "Invalid Curve pool");
+        require(
+            _curveConfig.curveLpToken != address(0) ||
+                Helpers.getDecimals(_curveConfig.curveLpToken) == 18,
+            "Invalid Curve LP token"
+        );
+
+        CURVE_BASE_ASSETS = _curveConfig.curveBaseAssets;
+        CURVE_POOL = _curveConfig.curvePool;
+        CURVE_LP_TOKEN = _curveConfig.curveLpToken;
+
+        address asset0 = ICurvePool(_curveConfig.curvePool).coins(0);
+        coin0 = asset0;
+        decimals0 = Helpers.getDecimals(asset0);
+
+        address asset1 = ICurvePool(_curveConfig.curvePool).coins(1);
+        coin1 = asset1;
+        decimals1 = Helpers.getDecimals(asset1);
+
+        // Only get the address of the third coin (index 2) if a three coin pool
+        address asset2 = _curveConfig.curveBaseAssets == 3
+            ? ICurvePool(_curveConfig.curvePool).coins(2)
+            : address(0);
+        coin2 = asset2;
+        decimals2 = _curveConfig.curveBaseAssets == 3
+            ? Helpers.getDecimals(asset2)
+            : 0;
```

Comment:
nit: Why not just use a `if` block instead of conditional statements? Would look a bit cleaner that way

Reply:
I agree its harder to read. Unfortunately, the version of Solidity we are using restricts "Immutable variables cannot be initialized inside an if statement."

The newer Solidity versions are less restrictive